### PR TITLE
Ordered starting of node components and clear dependencies between them

### PR DESCRIPTION
**Pull request by shamsasari**
Moved start up logic of the various node components out of their c’tors and into “start” methods, which are called from Node.start(). The components themselves are created in the Node’s c’tor with minimal initialisation logic.

Certain things are not immediately available at construction time, which are instead given to the components at start time in an orderly fashion:

* Certs from the node’s key store and trust store
* The network parameters, and thus access to things like maxMessageSize and the contracts whitelist
* A running database - several components were doing database stuff their c’tors
* The node’s NodeInfo, and thus access to things like identities and addresses

The messaging service couldn’t be created in the Node’s c’tor due to initialisation issues with MockNode. This should be fixed in a later commit.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 1]
```diff
@@ -169,198 +234,158 @@ abstract class AbstractNode(val configuration: NodeConfiguration,
...
+
+            val frozenTokenizableServices = tokenizableServices!!
+            tokenizableServices = null
+
+            verifyCheckpointsCompatible(frozenTokenizableServices)
+
+            smm.start(frozenTokenizableServices)
+            // Shut down the SMM so no Fibers are scheduled.
+            runOnStop += { smm.stop(acceptableLiveFiberCountOnStop()) }
+            (smm as? StateMachineManagerInternal)?.let {
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: can smm here ever _not_ be StateMachineManagerInternal?

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 2]
```diff
@@ -762,27 +738,25 @@ abstract class AbstractNode(val configuration: NodeConfiguration,
...
     class DatabaseConfigurationException(msg: String) : CordaException(msg)
 
-    protected open fun initialiseDatabasePersistence(schemaService: SchemaService,
-                                                     wellKnownPartyFromX500Name: (CordaX500Name) -> Party?,
-                                                     wellKnownPartyFromAnonymous: (AbstractParty) -> Party?): CordaPersistence {
+    protected open fun startDatabase() {
         val props = configuration.dataSourceProperties
         if (props.isEmpty) throw DatabaseConfigurationException("There must be a database configured.")
-        val database = configureDatabase(props, configuration.database, wellKnownPartyFromX500Name, wellKnownPartyFromAnonymous, schemaService)
+        database.hikariStart(props)
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: maybe `startHikariPool` instead

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 3]
```diff
@@ -164,67 +160,86 @@ open class Node(configuration: NodeConfiguration,
     //
     // The primary work done by the server thread is execution of flow logics, and related
     // serialisation/deserialisation work.
-    override lateinit var serverThread: AffinityExecutor.ServiceAffinityExecutor
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: I wonder if it's time to update the discussion above this line to reflect what we've done with enterprise.

File: node/src/main/kotlin/net/corda/node/services/keys/KeyManagementServiceInternal.kt [hunk 4]
```diff
@@ -0,0 +1,8 @@
+package net.corda.node.services.keys
+
+import net.corda.core.node.services.KeyManagementService
+import java.security.KeyPair
+
+interface KeyManagementServiceInternal : KeyManagementService {
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/services/keys/KeyManagementServiceInternal.kt]: Does this really justify its own file, and should it be in an internal package? I know all of node is not stabilised at the moment but in future we might want to expose parts of it.

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt [hunk 5]
```diff
@@ -76,10 +74,7 @@ class NetworkMapCacheImpl(
  * Extremely simple in-memory cache of the network map.
  */
 @ThreadSafe
-open class PersistentNetworkMapCache(
-        private val database: CordaPersistence,
-        notaries: List<NotaryInfo>
-) : SingletonSerializeAsToken(), NetworkMapCacheBaseInternal {
+open class PersistentNetworkMapCache(private val database: CordaPersistence) : SingletonSerializeAsToken(), NetworkMapCacheBaseInternal {
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: Can you fix the kdoc description of the class? It's not in memory or extremely simple anymore :)

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt [hunk 6]
```diff
@@ -68,7 +67,11 @@ class CordaPersistence(
 
     data class Boundary(val txId: UUID, val success: Boolean)
 
-    init {
+    private var _dataSource: DataSource? = null
```
[the remark this answers is not in the document]
Comment (shamsasari) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/persistence/CordaPersistence.kt]: Yes, but I want a reasonable error message if this is accessed before `start`.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 7]
```diff
@@ -115,50 +114,116 @@ abstract class AbstractNode(val configuration: NodeConfiguration,
...
+            schemaService
+    )
+    init {
+        // TODO Break cyclic dependency
+        identityService.database = database
+    }
+    private val persistentNetworkMapCache = PersistentNetworkMapCache(database)
+    val networkMapCache = NetworkMapCacheImpl(persistentNetworkMapCache, identityService, database).tokenize()
+    val checkpointStorage = DBCheckpointStorage()
+    @Suppress("LeakingThis")
```
[the remark this answers is not in the document]
Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: `makeTransactionStorage` and the like are open methods. If you're not careful in the sub-class you can access sub-class fields in them before they're initialised.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 8]
```diff
@@ -115,50 +114,116 @@ abstract class AbstractNode(val configuration: NodeConfiguration,
...
+    private val schedulerService = NodeSchedulerService(
+            platformClock,
+            database,
+            flowStarter,
+            servicesForResolution,
+            flowLogicRefFactory,
+            nodeProperties,
+            configuration.drainingModePollPeriod,
+            unfinishedSchedules = busyNodeLatch
+    ).tokenize().closeOnStop()
```
[the remark this answers is not in the document]
Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: I assume these have leaked in the past hence their presence in the old `makeServices` method which added them to be tokenised. I've tokenised the same services here.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 9]
```diff
@@ -169,198 +234,158 @@ abstract class AbstractNode(val configuration: NodeConfiguration,
...
-        val schemaService = NodeSchemaService(cordappLoader.cordappSchemas)
+        persistentNetworkMapCache.start(notaries = emptyList())
+        val trustRoot = initKeyStore()
         val (identity, identityKeyPair) = obtainIdentity(notaryConfig = null)
-        // Wrapped in an atomic reference just to allow setting it before the closure below gets invoked.
-        val identityServiceRef = AtomicReference<IdentityService>()
-        val database = initialiseDatabasePersistence(schemaService, { name -> identityServiceRef.get().wellKnownPartyFromX500Name(name) }, { party -> identityServiceRef.get().wellKnownPartyFromAnonymous(party) })
-        val identityService = makeIdentityService(identity.certificate, database)
-        identityServiceRef.set(identityService)
+        startDatabase()
```
[the remark this answers is not in the document]
Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Probably not.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt (hunk 1 above)
Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Good point.

File: node/src/main/kotlin/net/corda/node/services/keys/KeyManagementServiceInternal.kt (hunk 4 above)
Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/services/keys/KeyManagementServiceInternal.kt]: I could add it to KMSUtils.

File: node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt (hunk 5 above)
Comment (shamsasari) [reply node/src/main/kotlin/net/corda/node/services/network/PersistentNetworkMapCache.kt]: :)

**Comment by shamsasari** [comment]
I will address your comments in a separate PR, otherwise I'll lose the green tick!

**Pull request merged**